noir/aztec_macros/src/lib.rs

```diff
@@ -1601,3 +1601,143 @@ fn event_signature(event: &StructType) -> String {
+    let mut contract_module_file_ids: Vec<(LocalModuleId, FileId)> = context.def_map(crate_id).unwrap().modules().iter()
+        .filter(|(_, module)| module.is_contract)
+        .map(|(idx, module)| (LocalModuleId(idx), module.location.file))
+        .collect();
+
+    // If the current crate does not contain a contract module we simply skip it.
+    if contract_module_file_ids.len() == 0 {
+        return Ok(());
+    } else if contract_module_file_ids.len() != 1 {
+        panic!("Found multiple contracts in the same crate");
+    }
+
+    let (module_id, file_id) = contract_module_file_ids.pop().unwrap();
+
+    // If compute_note_hash_and_nullifier is already defined by the user, we skip autogeneration in order to provide an
+    // escape hatch for this mechanism.
+    if collected_functions.iter().any(|coll_funcs_data| check_for_compute_note_hash_and_nullifier_definition(&coll_funcs_data.functions, module_id)) {
+        return Ok(());
+    }
+
+    // In order to implement compute_note_hash_and_nullifier, we need to know all of the different note types the
+    // contract might use. These are the types that implement the NoteInterface trait, which provides the
+    // get_note_type_id function.
+
+    let mut note_types: Vec<String> = Vec::new();
+
+    // These note types can be declared in either external crates, or the contract's own crate. External crates have
+    // already been processed and resolved, but are available here via the NodeInterner. This is because the contract's
+    // crate is the last crate to be processed, after all of its dependencies.
+    for trait_impl_id in 0..(&context.def_interner.next_trait_impl_id()).0 {
+        let trait_impl = &context.def_interner.get_trait_implementation(TraitImplId(trait_impl_id));
+
+        if trait_impl.borrow().ident.0.contents == "NoteInterface" {
+            if let Type::Struct(s, _) = &trait_impl.borrow().typ {
+                note_types.push(s.borrow().name.0.contents.clone());
+            } else {
+                panic!("Found impl for NoteInterface on non-Struct");
+            }
+        }
+    }
+
+    // This crate's traits and impls have not yet been resolved, so we look for impls in unresolved_trait_impls.
+    note_types.extend(unresolved_traits_impls.iter()
+        .filter(|trait_impl| trait_impl.trait_path.segments.last().unwrap().0.contents == "NoteInterface")
+        .filter_map(|trait_impl| match &trait_impl.object_type.typ {
+            UnresolvedTypeData::Named(path, _, _) => Some(path.segments.last().unwrap().0.contents.clone()),
+            _ => None,
+        }));
+
+    // We can now generate a version of compute_note_hash_and_nullifier tailored for the contract in this crate.
+    let func = generate_compute_note_hash_and_nullifier(&note_types);
+
+    // And inject the newly created function into the contract.
+
+    // We don't have a reasonable location for the source code of this autogenerated function, so we simply pass an
```

Comment:
Could you link this TODO (https://github.com/AztecProtocol/aztec-packages/issues/4373)